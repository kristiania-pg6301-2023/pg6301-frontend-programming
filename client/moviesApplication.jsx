import { BrowserRouter, Link, Route, Routes } from "react-router-dom";
import React from "react";

function FrontPage() {
  return (
    <ul>
      <li>
        <Link to={"/movies"}>List movies</Link>
      </li>
      <li>
        <Link to={"/movies/new"}>Add a movie</Link>
      </li>
    </ul>
  );
}

function MoviesRoutes() {
  return (
    <Routes>
      <Route path={"/"} element={<FrontPage />} />
      <Route path={"/movies"} element={<h2>Listing of all movies</h2>} />
      <Route path={"/movies/new"} element={<h1>Add movie</h1>} />
      <Route path={"*"} element={<h1>Not Found</h1>} />
    </Routes>
  );
}

export function MoviesApplication() {
  return (
    <>
      <header>
        <h1>Kristiania movies</h1>
      </header>
      <nav>
        <Link to={"/"}>Front page</Link>
      </nav>
      <main>
        <MoviesRoutes />
      </main>
      <footer>By Johannes Brodwall with ðŸ’š</footer>
    </>
  );
}
