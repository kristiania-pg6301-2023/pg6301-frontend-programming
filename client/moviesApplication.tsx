import React, { Dispatch, SetStateAction, useEffect, useState } from "react";
import { Route, Routes, useNavigate } from "react-router-dom";

interface Movie {
  _id: string;
  title: string;
  plot: string;
  year: number;
}

interface MovieParameters {
  countries: string[];
  years: number[];
}

export interface MovieQuery {
  countries?: string;
  year?: number;
}

export function ListMovies({
  movies,
  parameters,
  setQuery,
}: {
  movies: Movie[];
  parameters: MovieParameters;
  query: MovieQuery;
  setQuery: Dispatch<SetStateAction<MovieQuery>>;
}) {
  return (
    <>
      <h2>All movies</h2>
      <div className={"query-filter"}>
        <div>
          <label>
            Country: <br />
            <select
              onChange={(e) =>
                setQuery((old) => ({ ...old, countries: e.target.value }))
              }
            >
              <option></option>
              {parameters.countries.map((c) => (
                <option key={c}>{c}</option>
              ))}
            </select>
          </label>
        </div>
        <div>
          <label>
            Year: <br />
            <select
              onChange={(e) =>
                setQuery((old) => ({ ...old, year: parseInt(e.target.value) }))
              }
            >
              <option></option>
              {parameters.years.map((y) => (
                <option key={y}>{y}</option>
              ))}
            </select>
          </label>
        </div>
      </div>
      {movies.map((m) => (
        <div key={m._id}>
          <h3>{m.title}</h3>
          <p>{m.plot}</p>
        </div>
      ))}
    </>
  );
}

function AddMovie({ onCreate }: { onCreate(movie: Omit<Movie, "_id">): void }) {
  const [title, setTitle] = useState("");
  const [year, setYear] = useState("");
  const [plot, setPlot] = useState("");

  async function handleSubmit() {
    await onCreate({ title, plot, year: parseInt(year) });
  }

  return (
    <form method="dialog" onSubmit={handleSubmit}>
      <h2>Movies</h2>
      <div>
        Title: <br />
        <input value={title} onChange={(e) => setTitle(e.target.value)} />
      </div>
      <div>
        Year: <br />
        <input value={year} onChange={(e) => setYear(e.target.value)} />
      </div>
      <div>
        Plot: <br />
        <textarea value={plot} onChange={(e) => setPlot(e.target.value)} />
      </div>
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}

export interface MoviesRoutesProps {
  fetchMovies(query: MovieQuery): Promise<Movie[]>;

  fetchParameters(): Promise<MovieParameters>;

  insertMovie(movie: Movie): Promise<void>;
}

export function MoviesRoutes({
  fetchMovies,
  fetchParameters,
  insertMovie,
}: MoviesRoutesProps) {
  const navigate = useNavigate();
  const [movies, setMovies] = useState<Movie[]>([]);
  const [query, setQuery] = useState<MovieQuery>({});
  const [parameters, setParameters] = useState<MovieParameters>({
    countries: [],
    years: [],
  });

  async function loadMovies() {
    setMovies(await fetchMovies(query));
    setParameters(await fetchParameters());
  }

  useEffect(() => {
    loadMovies();
  }, [query]);

  async function handleCreate(movie: Movie) {
    await insertMovie(movie);
    await loadMovies();
    navigate("/");
  }

  return (
    <Routes>
      <Route
        path={"/"}
        element={
          <ListMovies
            movies={movies}
            parameters={parameters}
            query={query}
            setQuery={setQuery}
          />
        }
      />
      <Route
        path={"/movies/new"}
        element={<AddMovie onCreate={handleCreate} />}
      />
      <Route path={"*"} element={<h2>Not Found</h2>} />
    </Routes>
  );
}
