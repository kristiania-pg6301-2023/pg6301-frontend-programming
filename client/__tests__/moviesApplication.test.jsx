import { MoviesApplication } from "../moviesApplication";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";

describe("movie application", () => {
  it("looks okay", () => {
    const component = renderer.create(
      <MemoryRouter>
        <MoviesApplication />
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
  });

  it("lists movies", () => {
    const component = renderer.create(
      <MemoryRouter initialEntries={["/movies"]}>
        <MoviesApplication />
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
    expect(component.root.findByType("h2").children.join(" ")).toBe(
      "Listing of all movies",
    );
  });
});
