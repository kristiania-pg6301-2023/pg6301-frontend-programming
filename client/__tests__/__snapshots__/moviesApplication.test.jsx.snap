// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movie application lists movies 1`] = `
[
  <header>
    <h1>
      Kristiania movies
    </h1>
  </header>,
  <nav>
    <a
      href="/"
      onClick={[Function]}
    >
      Front page
    </a>
  </nav>,
  <main>
    <h2>
      Listing of all movies
    </h2>
  </main>,
  <footer>
    By Johannes Brodwall with ðŸ’š
  </footer>,
]
`;

exports[`movie application looks okay 1`] = `
[
  <header>
    <h1>
      Kristiania movies
    </h1>
  </header>,
  <nav>
    <a
      href="/"
      onClick={[Function]}
    >
      Front page
    </a>
  </nav>,
  <main>
    <ul>
      <li>
        <a
          href="/movies"
          onClick={[Function]}
        >
          List movies
        </a>
      </li>
      <li>
        <a
          href="/movies/new"
          onClick={[Function]}
        >
          Add a movie
        </a>
      </li>
    </ul>
  </main>,
  <footer>
    By Johannes Brodwall with ðŸ’š
  </footer>,
]
`;
